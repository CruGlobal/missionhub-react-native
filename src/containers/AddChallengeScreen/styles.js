import { StyleSheet } from 'react-native';

import theme from '../../theme';

export default StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.white,
    flexDirection: 'column',
  },
  backButton: {
    fontSize: 16,
    color: theme.lightGrey,
  },
  textInput: {
    fontSize: 32,
    fontWeight: '300',
    lineHeight: 38,
    color: theme.lightGrey,
    borderBottomWidth: 0,
    marginTop: 80,
    marginBottom: 26,
    paddingHorizontal: 32,
  },
  dateWrap: {
    borderColor: theme.extraLightGrey,
    borderTopWidth: 1,
    borderBottomWidth: 1,
    paddingVertical: 16,
    paddingHorizontal: 32,
  },
  dateLabel: {
    fontSize: 10,
    color: theme.lightGrey,
  },
  dateInput: {
    fontSize: 16,
    lineHeight: 22,
    color: theme.lightGrey,
  },
});
