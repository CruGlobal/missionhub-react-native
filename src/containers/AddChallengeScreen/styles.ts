import { StyleSheet } from 'react-native';

import theme from '../../theme';

export default StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.white,
    flexDirection: 'column',
  },
  textInput: {
    fontSize: 24,
    fontWeight: '300',
    lineHeight: 30,
    color: theme.grey,
    borderBottomWidth: 0,
    marginBottom: 16,
    paddingHorizontal: 32,
  },
  dateWrap: {
    borderColor: theme.extraLightGrey,
    borderTopWidth: 1,
    borderBottomWidth: 1,
    paddingVertical: 16,
    paddingHorizontal: 32,
  },
  dateLabel: {
    fontSize: 12,
    color: '#939393',
  },
  dateInput: {
    fontSize: 16,
    lineHeight: 22,
    color: theme.challengeBlue,
  },
  detailWrap: {
    borderColor: theme.extraLightGrey,
    borderTopWidth: 1,
    paddingVertical: 16,
    paddingHorizontal: 32,
  },
  detailLabel: {
    fontSize: 12,
    color: '#939393',
  },
  detailInput: {
    fontSize: 16,
    lineHeight: 24,
    color: theme.grey,
    borderBottomWidth: 0,
    marginBottom: 26,
    paddingBottom: 70,
  },
  challengeImage: {
    position: 'absolute',
    zIndex: -1,
    bottom: 0,
    right: 0,
  },
  backButton: {
    alignItems: 'flex-end',
    alignSelf: 'flex-end',
  },
  disabledButton: {
    backgroundColor: theme.lightGrey,
  },
});
