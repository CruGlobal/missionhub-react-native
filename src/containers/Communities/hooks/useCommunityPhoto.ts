import { ImageSourcePropType } from 'react-native';

import { orgIsGlobal } from '../../../utils/common';
import GLOBAL_COMMUNITY_IMAGE from '../../../../assets/images/globalCommunityImage.png';

export const useCommunityPhoto = (
  communityId: string,
  communityPhotoUrl?: string | null,
): ImageSourcePropType => {
  if (communityPhotoUrl) {
    return { uri: communityPhotoUrl };
  } else if (orgIsGlobal({ id: communityId })) {
    return GLOBAL_COMMUNITY_IMAGE;
  } else {
    return {};
  }
};
