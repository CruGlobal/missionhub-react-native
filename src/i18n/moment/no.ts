export default {
  months: [
    'january',
    'february',
    'march',
    'april',
    'may',
    'june',
    'july',
    'august',
    'september',
    'october',
    'november',
    'december',
  ],
  monthsShort: [
    'jan',
    'feb',
    'mar',
    'apr',
    'may',
    'jun',
    'jul',
    'aug',
    'sep',
    'oct',
    'nov',
    'dec',
  ],
  monthsParseExact: true,
  weekdays: [
    'sunday',
    'monday',
    'tuesday',
    'wednesday',
    'thursday',
    'friday',
    'saturday',
  ],
  weekdaysShort: ['sun', 'mon', 'tue', 'wed', 'thu', 'fri', 'sat'],
  weekdaysMin: ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'],
  weekdaysParseExact: true,
  longDateFormat: {
    LT: 'HH:mm',
    LTS: 'HH:mm:ss',
    L: 'DD/MM/YYYY',
    LL: 'D MMMM YYYY',
    LLL: 'D MMMM YYYY HH:mm',
    LLLL: 'dddd D MMMM YYYY HH:mm',
  },
  calendar: {
    sameDay: '[Today at] LT',
    nextDay: '[Tomorrow at] LT',
    nextWeek: 'dddd [at] LT',
    lastDay: '[Yesterday at] LT',
    lastWeek: '[Last] dddd [at] LT',
    sameElse: 'L',
  },
  relativeTime: {
    future: 'in %s',
    past: '%s ago',
    s: 'a few seconds',
    ss: '%d seconds',
    m: 'one minute',
    mm: '%d minutes',
    h: 'one hour',
    hh: '%d hours',
    d: 'one day',
    dd: '%d days',
    M: 'one month',
    MM: '%d months',
    y: 'one year',
    yy: '%d years',
  },
  dayOfMonthOrdinalParse: /\d{1,2}(er|e)/,
  ordinal: function(number: number) {
    return number + (number === 1 ? 'er' : 'e');
  },
  meridiemParse: /PD|MD/,
  isPM: function(input: string) {
    return input.charAt(0) === 'M';
  },
  // In case the meridiem units are not separated around 12, then implement
  // this function (look at locale/id.js for an example).
  // meridiemHour : function (hour, meridiem) {
  //     return /* 0-23 hour, given meridiem token and hour 1-12 */ ;
  // },
  meridiem: function(hours: number) {
    return hours < 12 ? 'PD' : 'MD';
  },
  week: {
    dow: 0, // Monday is the first day of the week.
    doy: 6, // Used to determine first week of the year.
  },
};
