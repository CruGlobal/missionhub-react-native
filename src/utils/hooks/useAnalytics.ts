import { useEffect } from 'react';
import { useIsFocused } from 'react-navigation-hooks';
import { useDispatch, useSelector } from 'react-redux';

import { trackScreenChange, ScreenContext } from '../../actions/analytics';
import { DrawerState } from '../../reducers/drawer';

export enum ANALYTICS_SCREEN_TYPES {
  screen,
  screenWithDrawer,
  drawer,
}

interface UseAnalyticsParams {
  screenName: string | string[];
  screenType?: ANALYTICS_SCREEN_TYPES;
  screenContext?: Partial<ScreenContext>;
}

export const useAnalytics = ({
  screenName,
  screenType = ANALYTICS_SCREEN_TYPES.screen,
  screenContext,
}: UseAnalyticsParams) => {
  const dispatch = useDispatch();
  const isFocused = useIsFocused();
  const isDrawerOpen = useSelector(
    ({ drawer }: { drawer: DrawerState }) => drawer.isOpen,
  );

  const handleScreenChange = (
    name: string | string[],
    context?: Partial<ScreenContext>,
  ) => {
    dispatch(trackScreenChange(name, context));
  };

  //normally screens should only respond to focus events
  useEffect(() => {
    if (isFocused && screenType === ANALYTICS_SCREEN_TYPES.screen) {
      handleScreenChange(screenName, screenContext);
    }
  }, [isFocused]);

  //if it is a drawer, or screen with a drawer, it should respond to drawer events in addition to focus events
  useEffect(() => {
    if (isFocused) {
      if (screenType === ANALYTICS_SCREEN_TYPES.drawer && isDrawerOpen) {
        handleScreenChange(screenName, screenContext);
      } else if (
        screenType === ANALYTICS_SCREEN_TYPES.screenWithDrawer &&
        !isDrawerOpen
      ) {
        handleScreenChange(screenName, screenContext);
      }
    }
  }, [isFocused, isDrawerOpen]);

  return handleScreenChange;
};
