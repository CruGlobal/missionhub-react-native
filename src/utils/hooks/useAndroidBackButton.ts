import { BackHandler } from 'react-native';
import { useEffect } from 'react';
// eslint-disable-next-line import/named
import { NavigationEventSubscription } from 'react-navigation';
import { useNavigation } from 'react-navigation-hooks';

export const useAndroidBackButton = (
  enableBackButton = false,
  onBackPress?: () => void,
) => {
  const navigation = useNavigation();

  let willFocus: NavigationEventSubscription;
  let willBlur: NavigationEventSubscription;

  useEffect(() => {
    const handleBackPress = () => {
      if (enableBackButton) {
        if (onBackPress) {
          onBackPress();
        } else {
          return false;
        }
      }
      return true;
    };

    const addListeners = () => {
      willFocus = navigation.addListener('willFocus', () => {
        BackHandler.addEventListener('hardwareBackPress', handleBackPress);
      });
      willBlur = navigation.addListener('willBlur', () => {
        BackHandler.removeEventListener('hardwareBackPress', handleBackPress);
      });
    };
    const removeListeners = () => {
      willFocus && willFocus.remove();
      willBlur && willBlur.remove();
      BackHandler.removeEventListener('hardwareBackPress', handleBackPress);
    };

    addListeners();
    return removeListeners;
  }, [enableBackButton, onBackPress]);
};
