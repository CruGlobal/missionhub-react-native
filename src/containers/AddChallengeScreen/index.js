import React, { Component } from 'react';
import { connect } from 'react-redux';
import { View, Keyboard, Image } from 'react-native';
import PropTypes from 'prop-types';
import { translate } from 'react-i18next';

import CHALLENGE from '../../../assets/images/challenge_bullseye.png';
import { Button, Text, Flex, Input } from '../../components/common';
import DatePicker from '../../components/DatePicker';
import theme from '../../theme';
import { formatApiDate } from '../../utils/common';
import BackButton from '../BackButton';

import styles from './styles';

// TODO: Setup translations
@translate('addChallenge')
class AddChallengeScreen extends Component {
  constructor(props) {
    super(props);
    const { isEdit, challenge } = props;

    const date = isEdit ? challenge.end_date : '';
    this.state = {
      title: isEdit ? challenge.title : '',
      date,
      disableBtn: !isEdit,
    };

    this.today = new Date();
  }

  onChangeTitle = title => {
    this.setState({ title, disableBtn: !(title && this.state.date) });
  };

  onChangeDate = date => {
    if (!date) {
      this.setState({ date: '', disableBtn: false });
    } else {
      this.setState({ date, disableBtn: !this.state.title });
    }
  };

  saveChallenge = () => {
    Keyboard.dismiss();
    const { title, date } = this.state;
    const formattedTitle = (title || '').trim();
    if (!formattedTitle || !date) {
      return;
    }
    const challenge = {
      title: formattedTitle,
      date: formatApiDate(new Date(date)),
    };
    console.log('challenge', challenge);

    this.props.onComplete(challenge);
  };

  render() {
    const { t, isEdit } = this.props;
    const { disableBtn, title, date } = this.state;

    return (
      <View style={styles.container}>
        <Flex value={1} align="center" justify="center">
          <Image source={CHALLENGE} resizeMode="contain" />
          <Text type="header" style={styles.header}>
            {isEdit ? t('editHeader') : t('addHeader')}
          </Text>
        </Flex>

        <Flex value={1} style={styles.fieldWrap}>
          <Text style={styles.label}>{t('titleLabel')}</Text>
          <Input
            onChangeText={this.onChangeTitle}
            value={title}
            autoFocus={false}
            autoCorrect={true}
            selectionColor={theme.white}
            returnKeyType="next"
            blurOnSubmit={true}
            placeholder={t('titlePlaceholder')}
            placeholderTextColor={theme.white}
          />
          <Text style={styles.label}>{t('dateLabel')}</Text>
          <DatePicker
            date={date}
            mode="date"
            placeholder={t('datePlaceholder')}
            minDate={this.today}
            onDateChange={this.onChangeDate}
          />
        </Flex>

        <Flex value={1} align="stretch" justify="end">
          <Button
            disabled={disableBtn}
            type="secondary"
            onPress={this.saveChallenge}
            text={(isEdit ? t('save') : t('add')).toUpperCase()}
            style={styles.createButton}
          />
        </Flex>
        <BackButton customIcon="deleteIcon" absolute={true} />
      </View>
    );
  }
}

AddChallengeScreen.propTypes = {
  onComplete: PropTypes.func.isRequired,
  isEdit: PropTypes.bool,
  challenge: PropTypes.object,
};

const mapStateToProps = (reduxState, { navigation }) => ({
  ...(navigation.state.params || {}),
});

export default connect(mapStateToProps)(AddChallengeScreen);
export const ADD_CHALLENGE_SCREEN = 'nav/ADD_CHALLENGE';
