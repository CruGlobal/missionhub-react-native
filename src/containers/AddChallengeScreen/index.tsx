import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux-legacy';
import { View, Keyboard, StatusBar, ScrollView, Image } from 'react-native';
import { useTranslation } from 'react-i18next';
import { useNavigationParam } from 'react-navigation-hooks';
import moment from 'moment';

import { ANALYTICS_PERMISSION_TYPE } from '../../constants';
import { TrackStateContext } from '../../actions/analytics';
import { Text, Input, Button } from '../../components/common';
import DatePicker from '../../components/DatePicker';
import theme from '../../theme';
import DeprecatedBackButton from '../DeprecatedBackButton';
import BottomButton from '../../components/BottomButton';
import Header from '../../components/Header';
import { getAnalyticsPermissionType } from '../../utils/analytics';
import { AuthState } from '../../reducers/auth';
import CLOSE_BUTTON from '../../../assets/images/closeButton.png';
import CHALLENGE_TARGET from '../../../assets/images/challengeDetailsTarget.png';
import { useAnalytics } from '../../utils/hooks/useAnalytics';
import { isAndroid } from '../../utils/common';

import styles from './styles';

interface ChallengeInterface {
  id: string;
  title: string;
  created_at: string;
  end_date: string;
  details_markdown?: string;
}

interface AddChallengeScreenProps {
  analyticsPermissionType: TrackStateContext[typeof ANALYTICS_PERMISSION_TYPE];
}

const AddChallengeScreen = ({
  analyticsPermissionType,
}: AddChallengeScreenProps) => {
  const { t } = useTranslation('addChallenge');
  const isEdit: boolean = useNavigationParam('isEdit');
  const onComplete: (challenge: {
    title: string;
    date: string;
    id: string;
    details?: string;
  }) => void = useNavigationParam('onComplete');
  const challenge: ChallengeInterface = useNavigationParam('challenge');
  const [title, changeTitle] = useState(isEdit ? challenge.title : '');
  const [detail, changeDetail] = useState(
    isEdit ? challenge.details_markdown : '',
  );
  const [date, changeDate] = useState(
    isEdit ? moment(challenge.end_date).endOf('day') : '',
  );
  const [isEditing, setEditing] = useState('');
  const [disableBtn, changeDisableBtn] = useState(true);

  useEffect(() => {
    date !== '' && title !== ''
      ? changeDisableBtn(false)
      : changeDisableBtn(true);
  }, [date, title, detail]);

  useAnalytics(['challenge', isEdit ? 'edit' : 'create'], {
    screenContext: { [ANALYTICS_PERMISSION_TYPE]: analyticsPermissionType },
  });

  const {
    container,
    textInput,
    dateWrap,
    dateLabel,
    dateInput,
    detailWrap,
    detailLabel,
    detailInput,
  } = styles;

  const onChangeTitle = (challengeTitle: string) => {
    changeTitle(challengeTitle);
    changeDisableBtn(!(title && date));
  };

  const onChangeDate = (challengeDate: string) => {
    if (!challengeDate) {
      changeDate('');
      changeDisableBtn(false);
    } else {
      changeDate(challengeDate);
      changeDisableBtn(!title);
    }
  };

  const saveChallenge = () => {
    Keyboard.dismiss();
    const formattedTitle = (title || '').trim();
    if (!formattedTitle || !date) {
      return;
    }
    const newChallenge = {
      title: formattedTitle,
      // @ts-ignore
      date: moment(new Date(date))
        .endOf('day')
        .format(),
      id: '',
      details: detail,
    };
    if (isEdit) {
      newChallenge.id = challenge.id;
    }
    onComplete(newChallenge);
  };

  const renderTitleInput = () => {
    return (
      <Input
        testID="titleInput"
        onChangeText={onChangeTitle}
        onFocus={() => setEditing('title')}
        onBlur={() => setEditing('')}
        value={title}
        autoFocus={true}
        autoCorrect={true}
        multiline={true}
        returnKeyType="done"
        blurOnSubmit={true}
        selectionColor={theme.secondaryColor}
        placeholder={t(isEdit ? 'titlePlaceholderEdit' : 'titlePlaceholderAdd')}
        placeholderTextColor={
          isEditing === 'title' ? theme.lightGrey : theme.parakeetBlue
        }
        style={textInput}
      />
    );
  };

  const renderDateInput = () => {
    const today = new Date();
    return (
      <DatePicker
        // @ts-ignore
        date={date}
        mode="date"
        minDate={today}
        onDateChange={onChangeDate}
        testID="datePicker"
      >
        <View style={dateWrap}>
          <Text style={dateLabel}>{t('dateLabel')}</Text>
          <Text style={dateInput}>
            {!date ? t('datePlaceholder') : moment(date).format('LL')}
          </Text>
        </View>
      </DatePicker>
    );
  };

  const renderDetailInput = () => {
    return (
      <View style={detailWrap}>
        <Text style={detailLabel}>{t('detailsLabel')}</Text>
        <Input
          testID="detailInput"
          scrollEnabled={false}
          onChangeText={e => changeDetail(e)}
          onFocus={() => setEditing('detail')}
          onBlur={() => setEditing('')}
          value={detail}
          autoFocus={false}
          autoCorrect={true}
          multiline={true}
          selectionColor={theme.secondaryColor}
          placeholder={t('detailPlaceholder')}
          placeholderTextColor={
            isEditing === 'detail' ? theme.lightGrey : theme.parakeetBlue
          }
          style={detailInput}
        />
      </View>
    );
  };

  return (
    <View style={container}>
      <StatusBar {...theme.statusBar.darkContent} />
      <Header
        left={
          isEdit ? (
            <DeprecatedBackButton iconStyle={{ color: theme.lightGrey }} />
          ) : null
        }
        right={
          isEdit ? (
            <Button
              type="transparent"
              testID="editButton"
              text={t('save').toUpperCase()}
              onPress={saveChallenge}
              buttonTextStyle={{ color: theme.parakeetBlue, fontSize: 14 }}
              style={{ marginRight: 10 }}
            />
          ) : (
            <DeprecatedBackButton
              image={CLOSE_BUTTON}
              style={styles.backButton}
            />
          )
        }
      />
      <ScrollView
        contentContainerStyle={styles.fieldWrap}
        style={{ marginBottom: isAndroid ? 80 : undefined }}
        contentInset={{ bottom: 96 }}
      >
        {renderTitleInput()}
        {renderDateInput()}
        {renderDetailInput()}
      </ScrollView>
      <Image source={CHALLENGE_TARGET} style={styles.challengeImage} />
      {isEdit ? null : (
        <BottomButton
          testID="saveChallengeButton"
          style={disableBtn ? styles.disabledButton : null}
          disabled={disableBtn}
          onPress={saveChallenge}
          text={t('add')}
        />
      )}
    </View>
  );
};

const mapStateToProps = (
  { auth }: { auth: AuthState },
  {
    navigation,
  }: // eslint-disable-next-line @typescript-eslint/no-explicit-any
  any,
) => {
  const { organization } = navigation.state.params;

  return {
    ...(navigation.state.params || {}),
    analyticsPermissionType: getAnalyticsPermissionType(auth, organization),
  };
};

export default connect(mapStateToProps)(AddChallengeScreen);
export const ADD_CHALLENGE_SCREEN = 'nav/ADD_CHALLENGE';
