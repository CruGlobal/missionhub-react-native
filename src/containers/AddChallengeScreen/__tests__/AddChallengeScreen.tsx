import React from 'react';
import MockDate from 'mockdate';
import moment from 'moment';
import { fireEvent } from 'react-native-testing-library';

import { useAnalytics } from '../../../utils/hooks/useAnalytics';
import { renderWithContext } from '../../../../testUtils';

import AddChallengeScreen from '..';

const mockDate = '2020-02-13';
MockDate.set(mockDate);

jest.mock('../../../utils/hooks/useAnalytics');
jest.mock('../../../components/DatePicker', () => 'DatePicker');

const editChallenge = {
  id: '1',
  title: 'Test Title',
  end_date: '2020-02-29',
};
const onComplete = jest.fn();

it('renders correctly', () => {
  renderWithContext(<AddChallengeScreen />, {
    initialState: {},
    navParams: {
      onComplete,
      isEdit: false,
    },
  }).snapshot();
  expect(useAnalytics).toHaveBeenCalledWith(['challenge', 'create']);
});

it('renders edit challenge correctly', () => {
  renderWithContext(<AddChallengeScreen />, {
    initialState: {},
    navParams: {
      onComplete,
      challenge: editChallenge,
      isEdit: true,
    },
  }).snapshot();
  expect(useAnalytics).toHaveBeenCalledWith(['challenge', 'edit']);
});

describe('create methods', () => {
  it('focuses on press | titleInput', async () => {
    const { getByTestId, recordSnapshot, diffSnapshot } = renderWithContext(
      <AddChallengeScreen />,
      {
        initialState: {},
        navParams: {
          onComplete,
          isEdit: false,
        },
      },
    );
    recordSnapshot();
    await fireEvent(getByTestId('titleInput'), 'onFocus');
    diffSnapshot();
  });
  it('focuses on press | detailInput', async () => {
    const { getByTestId, recordSnapshot, diffSnapshot } = renderWithContext(
      <AddChallengeScreen />,
      {
        initialState: {},
        navParams: {
          onComplete,
          isEdit: false,
        },
      },
    );
    recordSnapshot();
    await fireEvent(getByTestId('detailInput'), 'onFocus');
    diffSnapshot();
  });

  it('fires onBlur | titleInput', async () => {
    const { getByTestId, recordSnapshot, diffSnapshot } = renderWithContext(
      <AddChallengeScreen />,
      {
        initialState: {},
        navParams: {
          onComplete,
          isEdit: false,
        },
      },
    );
    await fireEvent(getByTestId('titleInput'), 'onFocus');
    recordSnapshot();
    await fireEvent(getByTestId('detailInput'), 'onFocus');
    diffSnapshot();
  });

  it('fires onBlur | detailInput', async () => {
    const { getByTestId, recordSnapshot, diffSnapshot } = renderWithContext(
      <AddChallengeScreen />,
      {
        initialState: {},
        navParams: {
          onComplete,
          isEdit: false,
        },
      },
    );
    await fireEvent(getByTestId('detailInput'), 'onFocus');
    recordSnapshot();
    await fireEvent(getByTestId('titleInput'), 'onFocus');
    diffSnapshot();
  });

  it('changes title', async () => {
    const { getByTestId, recordSnapshot, diffSnapshot } = renderWithContext(
      <AddChallengeScreen />,
      {
        initialState: {},
        navParams: {
          onComplete,
          isEdit: false,
        },
      },
    );
    recordSnapshot();
    await fireEvent.changeText(
      getByTestId('titleInput'),
      'New Title For My Challenge',
    );
    diffSnapshot();
  });

  it('changes detail', async () => {
    const { getByTestId, recordSnapshot, diffSnapshot } = renderWithContext(
      <AddChallengeScreen />,
      {
        initialState: {},
        navParams: {
          onComplete,
          isEdit: false,
        },
      },
    );
    recordSnapshot();
    await fireEvent.changeText(
      getByTestId('detailInput'),
      'New details for my challenge',
    );
    diffSnapshot();
  });

  it('changes date', async () => {
    const date = new Date();
    const { getByTestId, recordSnapshot, diffSnapshot } = renderWithContext(
      <AddChallengeScreen />,
      {
        initialState: {},
        navParams: {
          onComplete,
          isEdit: false,
        },
      },
    );
    recordSnapshot();
    await fireEvent(getByTestId('datePicker'), 'onDateChange', date);
    diffSnapshot();
  });

  it('clears date', async () => {
    const date = new Date();
    const { getByTestId, recordSnapshot, diffSnapshot } = renderWithContext(
      <AddChallengeScreen />,
      {
        initialState: {},
        navParams: {
          onComplete,
          isEdit: false,
        },
      },
    );
    await fireEvent(getByTestId('datePicker'), 'onDateChange', date);
    recordSnapshot();
    await fireEvent(getByTestId('datePicker'), 'onDateChange', '');
    diffSnapshot();
  });

  it('sets disable false', async () => {
    const title = 'New Title';
    const date = new Date();
    const { getByTestId, recordSnapshot, diffSnapshot } = renderWithContext(
      <AddChallengeScreen />,
      {
        initialState: {},
        navParams: {
          onComplete,
          isEdit: false,
        },
      },
    );
    recordSnapshot();
    await fireEvent.changeText(getByTestId('titleInput'), title);
    await fireEvent(getByTestId('datePicker'), 'onDateChange', date);
    diffSnapshot();
  });

  it('calls onComplete', async () => {
    const title = 'New Title';
    const date = new Date();
    const details = 'New details for my challenge';
    const challenge = {
      title,
      date: moment(date)
        .endOf('day')
        .format(),
      id: '',
      details,
    };
    const { getByTestId, snapshot } = renderWithContext(
      <AddChallengeScreen />,
      {
        initialState: {},
        navParams: {
          onComplete,
          isEdit: false,
        },
      },
    );
    await fireEvent.changeText(getByTestId('titleInput'), title);
    await fireEvent(getByTestId('datePicker'), 'onDateChange', date);
    await fireEvent.changeText(getByTestId('detailInput'), details);
    await fireEvent.press(getByTestId('saveChallengeButton'));
    expect(onComplete).toHaveBeenCalledWith(challenge);
    snapshot();
  });
});
describe('edit methods', () => {
  const title = 'Old Title';
  const date = new Date();
  const details = 'New details for my challenge';
  const challenge = {
    title,
    date: moment(date)
      .endOf('day')
      .format(),
    id: '123',
  };
  it('calls onComplete', async () => {
    const { getByTestId, snapshot } = renderWithContext(
      <AddChallengeScreen />,
      {
        initialState: {},
        navParams: {
          onComplete,
          isEdit: true,
          challenge: { ...challenge, details_markdown: details },
        },
      },
    );
    await fireEvent.press(getByTestId('editButton'));
    expect(onComplete).toHaveBeenCalledWith({ ...challenge, details });
    snapshot();
  });

  it('disables button if user deletes title', async () => {
    const { getByTestId, recordSnapshot, diffSnapshot } = renderWithContext(
      <AddChallengeScreen />,
      {
        initialState: {},
        navParams: {
          onComplete,
          isEdit: true,
          challenge: { ...challenge, details_markdown: details },
        },
      },
    );
    recordSnapshot();
    await fireEvent.changeText(getByTestId('titleInput'), '');
    await fireEvent.press(getByTestId('editButton'));
    expect(onComplete).not.toHaveBeenCalled();
    diffSnapshot();
  });

  it('Saves challenge when user updates detail', async () => {
    const { getByTestId, recordSnapshot, diffSnapshot } = renderWithContext(
      <AddChallengeScreen />,
      {
        initialState: {},
        navParams: {
          onComplete,
          isEdit: true,
          challenge: { ...challenge, details_markdown: details },
        },
      },
    );
    recordSnapshot();
    await fireEvent.changeText(getByTestId('detailInput'), '');
    await fireEvent.press(getByTestId('editButton'));
    expect(onComplete).toHaveBeenCalledWith({ ...challenge, details: '' });
    diffSnapshot();
  });
});
