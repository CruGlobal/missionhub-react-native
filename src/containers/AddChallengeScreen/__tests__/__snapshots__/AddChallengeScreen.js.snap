// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <StatusBar
    animated={true}
    backgroundColor="#006789"
    barStyle="dark-content"
    showHideTransition="fade"
  />
  <View
    flex={1}
  >
    <Input
      autoCorrect={true}
      autoFocus={false}
      blurOnSubmit={true}
      multiline={true}
      onChangeText={[Function]}
      placeholder="New Challenge"
      placeholderTextColor="#B4B6BA"
      returnKeyType="done"
      selectionColor="#52C5DC"
      style={
        Object {
          "borderBottomWidth": 0,
          "color": "#B4B6BA",
          "fontSize": 32,
          "fontWeight": "300",
          "lineHeight": 38,
          "marginBottom": 26,
          "marginTop": 64,
          "paddingHorizontal": 32,
        }
      }
      value=""
    />
    <withI18nextTranslation(MyDatePickerIOS)
      customStyles={Object {}}
      date=""
      disabled={false}
      duration={300}
      height={259}
      minDate={2018-09-01T00:00:00.000Z}
      mode="date"
      onDateChange={[Function]}
    >
      <View
        style={
          Object {
            "borderBottomWidth": 1,
            "borderColor": "#ECEEF2",
            "borderTopWidth": 1,
            "paddingHorizontal": 32,
            "paddingVertical": 16,
          }
        }
      >
        <MyText
          style={
            Object {
              "color": "#B4B6BA",
              "fontSize": 10,
            }
          }
        >
          End Date
        </MyText>
        <MyText
          style={
            Object {
              "color": "#B4B6BA",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          End Date (Required)
        </MyText>
      </View>
    </withI18nextTranslation(MyDatePickerIOS)>
  </View>
  <BottomButton
    disabled={true}
    onPress={[Function]}
    text="Create Challenge"
  />
  <Connect(BackButton)
    absolute={true}
    iconStyle={
      Object {
        "color": "#B4B6BA",
        "fontSize": 16,
      }
    }
  />
</SafeAreaView>
`;

exports[`renders edit challenge correctly 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <StatusBar
    animated={true}
    backgroundColor="#006789"
    barStyle="dark-content"
    showHideTransition="fade"
  />
  <View
    flex={1}
  >
    <Input
      autoCorrect={true}
      autoFocus={false}
      blurOnSubmit={true}
      multiline={true}
      onChangeText={[Function]}
      placeholder="Edit Challenge"
      placeholderTextColor="#B4B6BA"
      returnKeyType="done"
      selectionColor="#52C5DC"
      style={
        Object {
          "borderBottomWidth": 0,
          "color": "#B4B6BA",
          "fontSize": 32,
          "fontWeight": "300",
          "lineHeight": 38,
          "marginBottom": 26,
          "marginTop": 64,
          "paddingHorizontal": 32,
        }
      }
      value="Test Title"
    />
    <withI18nextTranslation(MyDatePickerIOS)
      customStyles={Object {}}
      date="2018-09-30 23:59:59"
      disabled={false}
      duration={300}
      height={259}
      minDate={2018-09-01T00:00:00.000Z}
      mode="date"
      onDateChange={[Function]}
    >
      <View
        style={
          Object {
            "borderBottomWidth": 1,
            "borderColor": "#ECEEF2",
            "borderTopWidth": 1,
            "paddingHorizontal": 32,
            "paddingVertical": 16,
          }
        }
      >
        <MyText
          style={
            Object {
              "color": "#B4B6BA",
              "fontSize": 10,
            }
          }
        >
          End Date
        </MyText>
        <MyText
          style={
            Object {
              "color": "#B4B6BA",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          September 30, 2018
        </MyText>
      </View>
    </withI18nextTranslation(MyDatePickerIOS)>
  </View>
  <BottomButton
    disabled={true}
    onPress={[Function]}
    text="Save Changes"
  />
  <Connect(BackButton)
    absolute={true}
    iconStyle={
      Object {
        "color": "#B4B6BA",
        "fontSize": 16,
      }
    }
  />
</SafeAreaView>
`;
